File: HeThongDauGia/generate_graph.js
const fs = require('fs');
const path = require('path');

const extractResultsPath = 'E:/HeThongDauGia/HeThongDauGia/.understand-anything/tmp/ua-file-extract-results-2.json';
const inputDataPath = 'E:/HeThongDauGia/HeThongDauGia/.understand-anything/tmp/ua-file-analyzer-input-2.json';
const outputDir = 'E:/HeThongDauGia/HeThongDauGia/.understand-anything/intermediate';

const extractData = JSON.parse(fs.readFileSync(extractResultsPath, 'utf8'));
const inputData = JSON.parse(fs.readFileSync(inputDataPath, 'utf8'));

const batchImportData = inputData.batchImportData || {};

const nodes = [];
const edges = [];

function getVietnameseSummary(name, category, typeStr) {
    name = name.toLowerCase();
    if (name.includes('dao') || name.includes('repository')) {
        return `Lớp xử lý truy xuất cơ sở dữ liệu (Data Access Object) cho ${typeStr}. Cung cấp các phương thức tương tác với DB.`;
    }
    if (name.includes('dto') || name.includes('request') || name.includes('response')) {
        return `Đối tượng truyền tải dữ liệu (DTO) đóng gói thông tin cho ${typeStr}. Được sử dụng để giao tiếp giữa các tầng.`;
    }
    if (name.includes('controller') || name.includes('router') || name.includes('handler')) {
        return `Thành phần điều hướng và xử lý yêu cầu (Controller/Router) cho ${typeStr}. Tiếp nhận và phản hồi các request từ client.`;
    }
    if (name.includes('service') || name.includes('engine')) {
        return `Lớp dịch vụ (Service) chứa logic nghiệp vụ cốt lõi cho ${typeStr}.`;
    }
    if (name.includes('entity') || name.includes('model')) {
        return `Thực thể dữ liệu (Entity) đại diện cho ${typeStr} trong hệ thống. Ánh xạ trực tiếp với cấu trúc cơ sở dữ liệu.`;
    }
    if (name.includes('factory')) {
        return `Mẫu thiết kế Factory chịu trách nhiệm khởi tạo các đối tượng ${typeStr} dựa trên các điều kiện đầu vào.`;
    }
    if (name.includes('test')) {
        return `Lớp kiểm thử (Test) chứa các kịch bản kiểm tra tính đúng đắn cho ${typeStr}.`;
    }
    if (name.includes('util') || name.includes('context')) {
        return `Thành phần tiện ích hỗ trợ, cung cấp các hàm hoặc trạng thái dùng chung cho ${typeStr}.`;
    }
    return `Tệp mã nguồn cung cấp chức năng liên quan đến ${typeStr} trong hệ thống.`;
}

function getTags(name) {
    const tags = [];
    name = name.toLowerCase();
    if (name.includes('test')) tags.push('test');
    if (name.includes('dao')) tags.push('database');
    if (name.includes('dto') || name.includes('request') || name.includes('response')) tags.push('data-model', 'serialization');
    if (name.includes('controller') || name.includes('router') || name.includes('handler')) tags.push('api-handler', 'entry-point');
    if (name.includes('service') || name.includes('engine')) tags.push('service');
    if (name.includes('entity') || name.includes('model')) tags.push('data-model');
    if (name.includes('factory')) tags.push('factory');
    if (name.includes('util')) tags.push('utility');
    
    if (tags.length === 0) tags.push('component', 'utility');
    
    // Ensure we have at least 3 tags
    if (tags.length < 3) {
        if (!tags.includes('component')) tags.push('component');
        if (!tags.includes('backend') && tags.length < 3) tags.push('backend');
        if (!tags.includes('java') && tags.length < 3) tags.push('java');
    }
    return tags.slice(0, 5);
}

function getComplexity(nonEmptyLines) {
    if (nonEmptyLines < 50) return 'simple';
    if (nonEmptyLines < 200) return 'moderate';
    return 'complex';
}

const fileNodes = new Set();
const allFuncNodes = new Map(); // name -> id

// First pass: create file nodes
for (const file of extractData.results) {
    const p = file.path;
    const baseName = path.basename(p);
    const summary = getVietnameseSummary(baseName, file.fileCategory, baseName);
    
    nodes.push({
        id: `file:${p}`,
        type: 'file',
        name: baseName,
        filePath: p,
        summary: summary,
        tags: getTags(baseName),
        complexity: getComplexity(file.nonEmptyLines || file.totalLines),
        languageNotes: "Mã nguồn Java với thiết kế hướng đối tượng."
    });
    fileNodes.add(`file:${p}`);

    // Create class nodes
    if (file.classes) {
        for (const cls of file.classes) {
            const lineCount = (cls.endLine - cls.startLine) + 1;
            const methodCount = cls.methods ? cls.methods.length : 0;
            const isExported = file.exports && file.exports.some(e => e.name === cls.name);
            
            if (lineCount >= 20 || methodCount >= 2 || isExported) {
                const clsId = `class:${p}:${cls.name}`;
                nodes.push({
                    id: clsId,
                    type: 'class',
                    name: cls.name,
                    filePath: p,
                    summary: `Lớp ${cls.name} định nghĩa cấu trúc và hành vi cho ${cls.name} trong hệ thống.`,
                    tags: getTags(cls.name),
                    complexity: getComplexity(lineCount),
                    lineRange: [cls.startLine, cls.endLine]
                });
                edges.push({
                    source: `file:${p}`,
                    target: clsId,
                    type: 'contains',
                    direction: 'forward',
                    weight: 1.0
                });
                if (isExported) {
                    edges.push({
                        source: `file:${p}`,
                        target: clsId,
                        type: 'exports',
                        direction: 'forward',
                        weight: 0.8
                    });
                }
            }
        }
    }

    // Create function nodes
    if (file.functions) {
        for (const func of file.functions) {
            const lineCount = (func.endLine - func.startLine) + 1;
            const isExported = file.exports && file.exports.some(e => e.name === func.name);
            
            if (lineCount >= 10 || isExported) {
                const funcId = `function:${p}:${func.name}`;
                allFuncNodes.set(func.name, funcId);
                nodes.push({
                    id: funcId,
                    type: 'function',
                    name: func.name,
                    filePath: p,
                    summary: `Hàm ${func.name} thực thi logic nghiệp vụ cụ thể cho ${baseName}.`,
                    tags: getTags(func.name),
                    complexity: getComplexity(lineCount),
                    lineRange: [func.startLine, func.endLine]
                });
                edges.push({
                    source: `file:${p}`,
                    target: funcId,
                    type: 'contains',
                    direction: 'forward',
                    weight: 1.0
                });
                if (isExported) {
                    edges.push({
                        source: `file:${p}`,
                        target: funcId,
                        type: 'exports',
                        direction: 'forward',
                        weight: 0.8
                    });
                }
            }
        }
    }

    // Imports edges
    const imports = batchImportData[p] || [];
    for (const imp of imports) {
        edges.push({
            source: `file:${p}`,
            target: `file:${imp}`,
            type: 'imports',
            direction: 'forward',
            weight: 0.7
        });
    }
    
    // tested_by edge heuristic
    if (p.includes('Test.java') || p.includes('/test/')) {
        let prodPath = p.replace('src/test/', 'src/main/').replace('Test.java', '.java');
        edges.push({
            source: `file:${p}`,
            target: `file:${prodPath}`,
            type: 'tested_by',
            direction: 'forward',
            weight: 0.5
        });
    }
}

// Split logic
const maxNodes = 60;
const maxEdges = 120;
const parts = Math.ceil(Math.max(nodes.length / maxNodes, edges.length / maxEdges));

if (parts <= 1) {
    const outPath = path.join(outputDir, `batch-2.json`);
    fs.writeFileSync(outPath, JSON.stringify({ nodes, edges }, null, 2));
    console.log(`Wrote 1 part to ${outPath}. Total nodes: ${nodes.length}, Total edges: ${edges.length}`);
} else {
    // Sort files alphabetically
    const files = extractData.results.map(f => f.path).sort();
    
    const chunkSize = Math.ceil(files.length / parts);
    for (let i = 0; i < parts; i++) {
        const chunkFiles = new Set(files.slice(i * chunkSize, (i + 1) * chunkSize));
        
        const partNodes = nodes.filter(n => chunkFiles.has(n.filePath));
        const partNodeIds = new Set(partNodes.map(n => n.id));
        
        const partEdges = edges.filter(e => {
            // Include edge if its source is in this part's nodes
            return partNodeIds.has(e.source) || chunkFiles.has(e.source.replace('file:', ''));
        const refinedEdges = edges.filter(e => {
            const parts = e.source.split(':');
            const sourcePath = parts.length >= 2 ? parts[1] : '';
            return chunkFiles.has(sourcePath);
        });

        const outPath = path.join(outputDir, `batch-2-part-${i + 1}.json`);
        fs.writeFileSync(outPath, JSON.stringify({ nodes: partNodes, edges: refinedEdges }, null, 2));
        console.log(`Wrote part ${i + 1} to ${outPath}. Nodes: ${partNodes.length}, Edges: ${refinedEdges.length}`);
    }
    console.log(`Split into ${parts} parts. Total nodes: ${nodes.length}, Total edges: ${edges.length}`);
}

File: HeThongDauGia/generate_graph.py
import json
import os
import math

def get_summary(filename):
    if "DAO.java" in filename:
        return "Lớp DAO (Data Access Object) quản lý thao tác với cơ sở dữ liệu."
    elif "Config.java" in filename:
        return "Lớp thực thể đại diện cho cấu hình dữ liệu."
    elif "Exception.java" in filename:
        return f"Lớp ngoại lệ (Exception) xử lý lỗi nghiệp vụ."
    elif "Service.java" in filename:
        return "Lớp dịch vụ (Service) trung tâm chứa logic nghiệp vụ."
    elif "Test.java" in filename:
        return "Lớp kiểm thử (Test) đảm bảo tính đúng đắn của hệ thống."
    else:
        return "Tệp mã nguồn Java."

def get_tags(filename):
    if "DAO.java" in filename:
        return ["truy-cap-du-lieu", "dao", "co-so-du-lieu"]
    elif "Config.java" in filename:
        return ["thuc-the", "mo-hinh-du-lieu", "cau-hinh"]
    elif "Exception.java" in filename:
        return ["xu-ly-ngoai-le", "loi", "nghiep-vu"]
    elif "Service.java" in filename:
        return ["dich-vu", "nghiep-vu", "xu-ly-logic", "core"]
    elif "Test.java" in filename:
        return ["kiem-thu", "unit-test", "kiem-tra"]
    else:
        return ["ma-nguon", "thanh-phan"]

def get_complexity(lines):
    if lines < 50: return "simple"
    elif lines <= 200: return "moderate"
    else: return "complex"

def main():
    with open('.understand-anything/tmp/ua-file-extract-results-1.json', 'r', encoding='utf-8') as f:
        extract_data = json.load(f)
        
    with open('.understand-anything/tmp/ua-file-analyzer-input-1.json', 'r', encoding='utf-8') as f:
        input_data = json.load(f)
        
    import_map = input_data.get('batchImportData', {})
    
    nodes = []
    edges = []
    batch_file_paths = set([r['path'] for r in extract_data.get('results', [])])
    
    for result in extract_data.get('results', []):
        path = result['path']
        lines = result.get('nonEmptyLines', 0)
        
        file_node_id = f"file:{path}"
        
        nodes.append({
            "id": file_node_id,
            "type": "file",
            "name": os.path.basename(path),
            "filePath": path,
            "summary": get_summary(path),
            "tags": get_tags(path),
            "complexity": get_complexity(lines),
            "languageNotes": "Mã nguồn Java." if "Exception" not in path else "Lớp ngoại lệ Java."
        })
        
        exports = [e['name'] for e in result.get('exports', [])]
        
        for func in result.get('functions', []):
            f_lines = func['endLine'] - func['startLine'] + 1
            is_exported = func['name'] in exports
            if f_lines >= 10 or is_exported:
                func_id = f"function:{path}:{func['name']}"
                nodes.append({
                    "id": func_id,
                    "type": "function",
                    "name": func['name'],
                    "summary": f"Phương thức {func['name']} xử lý logic.",
                    "tags": ["phuong-thuc", "logic"],
                    "complexity": get_complexity(f_lines),
                    "lineRange": [func['startLine'], func['endLine']]
                })
                edges.append({
                    "source": file_node_id,
                    "target": func_id,
                    "type": "contains",
                    "direction": "forward",
                    "weight": 1.0
                })
                if is_exported:
                    edges.append({
                        "source": file_node_id,
                        "target": func_id,
                        "type": "exports",
                        "direction": "forward",
                        "weight": 0.8
                    })
                    
        for cls in result.get('classes', []):
            c_lines = cls['endLine'] - cls['startLine'] + 1
            c_methods = len(cls.get('methods', []))
            is_exported = cls['name'] in exports
            if c_lines >= 20 or c_methods >= 2 or is_exported:
                cls_id = f"class:{path}:{cls['name']}"
                nodes.append({
                    "id": cls_id,
                    "type": "class",
                    "name": cls['name'],
                    "summary": f"Lớp {cls['name']}.",
                    "tags": ["lop", "huong-doi-tuong"],
                    "complexity": get_complexity(c_lines),
                    "lineRange": [cls['startLine'], cls['endLine']]
                })
                edges.append({
                    "source": file_node_id,
                    "target": cls_id,
                    "type": "contains",
                    "direction": "forward",
                    "weight": 1.0
                })
                if is_exported:
                    edges.append({
                        "source": file_node_id,
                        "target": cls_id,
                        "type": "exports",
                        "direction": "forward",
                        "weight": 0.8
                    })
                    
        imports = import_map.get(path, [])
        for imp in imports:
            edges.append({
                "source": file_node_id,
                "target": f"file:{imp}",
                "type": "imports",
                "direction": "forward",
                "weight": 0.7
            })
            if "Test.java" in path and "Test.java" not in imp:
                edges.append({
                    "source": f"file:{imp}",
                    "target": file_node_id,
                    "type": "tested_by",
                    "direction": "forward",
                    "weight": 0.5
                })

    node_count = len(nodes)
    edge_count = len(edges)
    
    parts = int(math.ceil(max(node_count / 60.0, edge_count / 120.0)))
    
    # Check max chunk limits manually to adjust `parts` if necessary.
    # The heuristic formula sometimes underestimates if edges are clustered.
    parts = max(parts, 3) 
        
    os.makedirs('.understand-anything/intermediate', exist_ok=True)
    
    file_paths = sorted(list(batch_file_paths))
    chunk_size = int(math.ceil(len(file_paths) / parts))
    
    # Erase old parts to prevent duplicate leftovers
    for f in os.listdir('.understand-anything/intermediate'):
        if f.startswith('batch-1-part-'):
            os.remove(os.path.join('.understand-anything/intermediate', f))
    
    for k in range(parts):
        chunk_files = set(file_paths[k * chunk_size : (k + 1) * chunk_size])
        if not chunk_files:
            continue
            
        part_nodes = [n for n in nodes if n.get('filePath') in chunk_files or n['id'].split(':')[1] in chunk_files]
        part_node_ids = set(n['id'] for n in part_nodes)
        
        part_edges = []
        for e in edges:
            src_path = e['source'].split(':')[1]
            tgt_path = e['target'].split(':')[1]
            
            if src_path in chunk_files:
                part_edges.append(e)
            elif src_path not in batch_file_paths and tgt_path in chunk_files:
                part_edges.append(e)
        
        with open(f'.understand-anything/intermediate/batch-1-part-{k+1}.json', 'w', encoding='utf-8') as f:
            json.dump({"nodes": part_nodes, "edges": part_edges}, f, indent=2, ensure_ascii=False)
        print(f"Wrote part {k+1}: batch-1-part-{k+1}.json with {len(part_nodes)} nodes and {len(part_edges)} edges.")

if __name__ == '__main__':
    main()
